src/core/utils/props.ts

```diff
@@ -26,3 +28,40 @@ export type CommonProps = {
    */
   id?: string;
 } & StylingProps;
+
+/**
+ * Merges provided Props with the props of T.
+ *
+ * T can be any native HTML element or a custom component.
+ */
+export type PolymorphicComponentProps<
+  T extends React.ElementType,
+  Props = Record<string, unknown>
+> = Merge<React.ComponentPropsWithoutRef<T>, Props>;
+
+/**
+ * Merges original OwnProps and the inferred props from `as` element.
+ *
+ * T should be the default element that is used for the `as`  prop.
+ */
+export interface PolymorphicForwardRefComponent<
```

Comment:
Any chance you can explain all this magic?

Reply:
First, we extend `React.ForwardRefExoticComponent` so that ref gets correct type. We also add the `as` prop here.

After that we progressively narrow down the type using chained ternaries, checking whether it's extending intrinsic element vs component and merging props accordingly.

Reply:
Maybe add this explanation as a comment. Also thinking maybe it is worth to add an example how to use these types but this one is up to you.

Reply:
Updated jsdoc